### PR TITLE
Add ExprSplitter

### PR DESCRIPTION
**Pull request by timholy**
This implements an "expression-splitting iterator" that returns individual expressions ready for evaluation to the caller. This allows the caller to decide what to do with them (e.g., evaluation, signature analysis, whatever). It's basically a proposed replacement for `prepare_thunk` and `split_expressions`; by placing the consumer in charge, we get rid of any of the weird "should I eval if I can't lower?" questions that have dogged the previous framework.

~~For the moment this is parallel to our older framework; I've not yet actually rebased JuliaInterpreter on this, although the `Frame(::Module, ::CodeInfo)` constructor is a step in this direction.~~ My local copy of `Revise#revise3` is now based on this and it seems to work well.

@pfitzseb and @aviatesk, are there callers of these functions outside JuliaInterpreter, Debugger, and Revise? If so we'd want to think about a careful deprecation path. Indeed we might want to do that anyway since `split_expressions`, even though not exported, is pretty essential in some contexts (so there may be unknown users).

Closes #406
Closes #419

### CODE REVIEW

**Comment by pfitzseb** [comment]
I'm pretty sure we're using `split_expressions` in VSCode and Juno, yes.

**Comment by KristofferC** [comment]
Instead of working hard on deprecation and then just have everyone immediately update to the new one (effectively making the work on the deprecation pointless) we could also release a new breaking version and have people just update to that. Of course, this depends on how hard it is to do the deprecation and how hard it is to do the upgrade.

**Comment by timholy** [comment]
Since these are officially internal (though well-used) functions, I decided to go with the breaking change. The transition is pretty straightforward as can be seen by the tests. The main thing I'm unsure of is whether we want to export a utility for detecting and stripping out `@doc` exprs, or whether that was relevant only for Revise.

File: test/interpret.jl
Comment (timholy) [inline test/interpret.jl:325]: I wasn't quite sure what to do about these.

**Comment by timholy** [comment]
Ugh, this is ugly getting it over the finish line. The various bugs that different Julia versions have wrt LineNumberNodes are really annoying.

Are the remaining errors anything to worry about? Travis had a failure on Windows 1.4 (a strange string-indexing error), appveyor on 1.2 (a problem with `stat`). The second seems more worrisome, since it seems to have happened on both 32- and 64-bit. A few older PRs were not triggering this.

File: src/utils.jl
Comment (aviatesk) [inline src/utils.jl:141]: It feels like `must_be_toplevel` actually means `must_be_just_evaluated` or something ?
Because it seems like what should always be in toplevel is actually only `:const`, and the others can work within local scopes (like `let`), like:
```julia
julia> let 
           i = 0
           struct A
               val
               A(v = i) = new(v)
           end
           export A
       end

julia> A()
A(0)
```

**Comment by aviatesk** [comment]
> Since these are officially internal (though well-used) functions, I decided to go with the breaking change. The transition is pretty straightforward as can be seen by the tests. The main thing I'm unsure of is whether we want to export a utility for detecting and stripping out @doc exprs, or whether that was relevant only for Revise.

as far as I know, the only consumer who uses the utility for `at-doc` exprs is [only Revise](https://juliahub.com/ui/CodeSearch?q=split_expressions&u=all&t=all), so I would like to say it's your call :)

File: src/utils.jl
Comment (timholy) [reply src/utils.jl:143]: *early* expansion (like in the first lines of the `ExprSplitter` constructor) triggers failures in [this test](https://github.com/JuliaDebug/JuliaInterpreter.jl/blob/64d719fd202fedd41d9b0400340708187dd38f8e/test/toplevel.jl#L249-L289), which is explicitly designed to catch flaws like that. But "last minute" expansion, in `queuenext!`, might be possible. But of course the user can do it too: if the returned expr has a `:macroexpand` then expand and re-queue. I was taking the viewpoint of "do no harm" and so didn't want to transform expressions more than necessary.

File: src/utils.jl
Comment (timholy) [reply src/utils.jl:141]: Yeah, in a way I guess it's really `doesnt_lower`.

**Comment by timholy** [comment]
Thanks! If Revise is the only one, I've already got that covered from within Revise. There's isn't much to it, really, just grabbing the 4th argument of the expression so it wouldn't be much of a utiliy anyway.

**Comment by KristofferC** [comment]
> Are the remaining errors anything to worry about? Travis had a failure on Windows 1.4 (a strange string-indexing error), appveyor on 1.2 (a problem with stat)

I would not really worry about unmaintained julia versions (1.1 - 1.4). Might be time to drop compat from them completely so they will keep working at the state they are in but not get updates.

**Comment by timholy** [comment]
Sounds good. Looks like https://github.com/JuliaLang/julia/issues/26685. I just pushed a debug commit; let's see if that reveals an easy fix, and if not I'll just make 1.3 the lowest Julia version supported.

**Comment by timholy** [comment]
That is one weird filename :smile:.https://ci.appveyor.com/project/KristofferC/juliainterpreter-jl/builds/35055810/job/ubmbpu36otugvd9e

**Comment by timholy** [comment]
A JuliaHub search indicates that the only known users of `prepare_thunk` and `split_expressions` are the ones we know about. So this crew (I should have also CCed @davidanthoff) can decide.

File: src/construct.jl
Comment (KristofferC) [inline src/construct.jl:516]: Should this be uncommented or removed?

**Comment by KristofferC** [comment]
> That is one weird filename

Wow, where did that come from?

File: src/construct.jl
Comment (timholy) [reply src/construct.jl:516]: It could be useful for future debugging, but I will remove it. Frequently when I develop code I first write a brittle implementation and then remove the error condition when I become convinced I've handled the necessary cases. But this will throw on many things, so there is little point any more.

Comment (KristofferC) [reply src/construct.jl:516]: > Frequently when I develop code I first write a brittle implementation and then remove the error condition when I become convinced I've handled the necessary cases. 

Yeah me too, but without any comment or such it might become a point of uncertainty for a future reader.

File: src/utils.jl
Comment (timholy) [reply src/utils.jl]: Thanks for pushing on this. It turns out that if you wrap in a `:block` together with a `LineNumberNode`, then even `export`, `import`, and `using` do lower to a `CodeInfo`. And `:const` already did, I'm not sure why I added it here.

That leaves only `:global` expressions. Does anyone know how to make them lower to a CodeInfo?

Given that it's just `:global`, I've reworked this to simply not deliver a `:block`/`:toplevel` wrapper when it's a `:global` expr.

**Comment by timholy** [comment]
I am not sure. Seems likely to be a memory bug or something.

**Comment by timholy** [comment]
`Revise@3.0.0` is officially ready to be announced, pending the resolution of this PR.

File: src/utils.jl
Comment (aviatesk) [reply src/utils.jl]: > And `:const` already did

I never know that, so sneaky.

> Given that it's just `:global`, I've reworked this to simply not deliver a `:block`/`:toplevel` wrapper when it's a `:global` expr.

Yeah, the docstring is kind enough, I think.

File: src/utils.jl
Comment (aviatesk) [reply src/utils.jl]: I see. Then let's go this way.

**Review by aviatesk — APPROVED** [review]
lgtm

**Comment by timholy** [comment]
Seems like folks are willing to do the work of updating for the breaking change. Many thanks, all!

**Comment by timholy** [comment]
Before merging I'm going to back out the commit disabling the ambiguity test. That means the test will fail, but we've at least seen the tests pass.

**Pull request merged**
